bob-backend/src/pearl/disk_controller.rs

```diff
@@ -465,15 +467,20 @@ impl DiskController {
 
     pub(crate) async fn get_alien(&self, op: Operation, key: BobKey) -> Result<BobData, Error> {
         if *self.state.read().await == GroupsState::Ready {
-            let vdisk_group = self.find_group(&op).await;
-            if let Ok(group) = vdisk_group {
-                group.get(key).await
+            let result = self.get_alien_results(&op, |g| async move { g.get(key).await },
```

Comment:
It seems like if you implement loop over groups right here (remove `get_alien_results` function) then the code will be much simpler.
```Rust
let mut result = None;
for g in self.find_all_groups(op).await {
    match g.get(key).await {
          Ok(data) if data.meta().timestamp() > result.map_or(0, |d| d.meta().timestamp()) => { 
              result = data; 
          },
          Ok(_) => { },
          Err(e) if e.is_key_not_found() => { },
          Err(e) => { 
              debug!("error getting data from alien for op {:?}: {:?}", op, e);
          }
     }
}
```

Reply:
IMHO, a regular loop in place has much better readability than calling a function that performs folding.

In any case, I looked at the code of `delete_alien`: the error handling logic should be different there. `delete_alien` should return error if any deletion also ended with error (it should process all groups, store the latest error and return it)